docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+
+- **OID:** Object Identifier defined by the International Telecommunications Union and used in PKI to identify attributes on certificates.
+
+- **PKI:** Public Key Infrastructure
+
+- **TPM:** Trusted Platform Module (TPM) is a specification defined by the Trusted Computing Group (TCG) that allows hosts to attest to their identity via PKI and a secure crypto-processor which may either be a separate chip, built into the CPU or a virtual device provided by the hypervisor.
+
+
+## Summary
+
+This proposal outlines a method to secure node registration within Cluster API, to solve 3 primary problems:
+- Solving a class of attacks involving node impersonation allowing an attacker to access secrets and volumes they shouldn’t by using hardware attestation of node identity.
+- Prevent token reuse in MachinePools where the token cannot be easily rotated.
+
+
+## Motivation
+
+Cluster API default core components are largely reliant on kubeadm for cluster bootstrapping and node registration. Kubeadm is a platform-agnostic command line tool designed to assist users to bootstrap Kubernetes clusters, and is used as a building block in Cluster API.
+Because kubeadm is platform-independent and is intended to provide an “easy path” to cluster bootstrapping,  there are a number of inherent design decisions that limit the overall security of the provisioned cluster:
+- Kubeadm uses TLS bootstrapping for node registration, however the default workflow used by Cluster API uses bootstrap token which allow registration as arbitrary node names.
+    - When used in this mode, Kubeadm essentially does “client-side validation” to prevent node hijacking, but this does not mean the token cannot be reused by an attacker within the lifetime of the token to perform a hijack.
+    - Cluster API cannot scope a token down  to a specific node, because bootstrap providers, nor most infrastructure providers know the identity of the node ahead of time.
+- Kubeadm sets node labels during bootstrapping, and Cluster API does not do reconciliation of node labels post node deployment.
+    - This is helpful because it means the kubelet starts with the labels applied straightaway, preventing a race condition when the Kubernetes schedule may misplace a pod due to a temporarily missing node label.
+    - However, again, this allows for potential hijacking, for example, using the kubelet’s credentials to apply a label intended for workloads not using the kubernetes.io/node-restriction-* labels.
+
+
+### Goals
+
+- Provide a bootstrap mechanism that assures node secure node registration
+- The node registration 
+
+
+### Non-Goals/Future Work
```

Comment:
can we also add the scope of this proposal in terms of providers? Is it expected to work with all infra providers? all bootstrap providers or just kubeadm?

Reply:
`node-attestation-controller` isn't depending on kubeadm itself. So the controller can be re-used. As for the scope, yes I'd expect this to be running for all infra providers if/when using #4221

Reply:
added as a goal that it's expected to work with any provider (i.e. status quo like in v1a3 and enhanced features for the providers who are able to support it)